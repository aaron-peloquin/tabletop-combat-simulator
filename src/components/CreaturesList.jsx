import { connect } from "react-redux"
import Link from "next/link"
import {
  Table,
  TableHead, TableBody,
  TableRow, TableCell,
  TableSortLabel
} from "@material-ui/core/"

const CreatureRow = ({data}) => {
  return <TableRow key={data.hash}>
    <TableCell>
      <Link as={`/creature/${data.hash}`} href={`/creature?hash=${data.hash}`}><a>Edit</a></Link>
      {data.name}
    </TableCell>
    <TableCell>1</TableCell>
    <TableCell>{data.description}</TableCell>
  </TableRow>
}

const CreaturesTable = ({data}) => {
  if(data.length>0) {
    const columns = ["Name","CR","Description"]
    return <Table>
      <TableHead>
        <TableRow>
          {columns.map((label)=>{return <TableCell key={label}>{label}</TableCell>})}
        </TableRow>
      </TableHead>
      <TableBody>
        {data.map((d)=><CreatureRow key={d.hash} data={d} />)}
      </TableBody>
    </Table>
  }
  return <p>No creatures.</p>
}

const CreaturesList = (props) => {
  const {creatures,numCreatures} = props
  return <div>
    <p>Showing {numCreatures} Creatures</p>
    <CreaturesTable data={creatures} />
  </div>
}

const mapStateToProps = (state) => {
  const creatures = state.creatures
  const numCreatures = state.creatures.length
  return {creatures, numCreatures}
}

export default connect(mapStateToProps)(CreaturesList)