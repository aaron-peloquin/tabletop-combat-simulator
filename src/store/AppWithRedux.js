import React from 'react'
import {initializeStore} from './store'

const isServer = typeof window === 'undefined'
const reduxStoreSymbol = Symbol("reduxStore")

function getOrCreateStore (initialState) {
  // Always make a new store if server, otherwise state is shared between requests
  if (isServer) {
    return initializeStore(initialState)
  }

  // Create store if unavailable on the client and set it on the window object
  if (!window[reduxStoreSymbol]) {
    window[reduxStoreSymbol] = initializeStore(initialState)
  }
  return window[reduxStoreSymbol]
}

export default (App) => {
  return class AppWithRedux extends React.Component {
    static async getInitialProps (appContext) {
      // Get or Create the store with `undefined` as initialState
      // This allows you to set a custom default initialState
      const reduxStore = getOrCreateStore()

      // Provide the store to getInitialProps of pages
      appContext.ctx.reduxStore = reduxStore

      let appProps = {}
      if (typeof App.getInitialProps === 'function') {
        appProps = await App.getInitialProps(appContext)
      }

      return {
        ...appProps,
        initialReduxState: reduxStore.getState()
      }
    }

    constructor (props) {
      super(props)
      this.reduxStore = getOrCreateStore(props.initialReduxState)
    }

    render () {
      return <App {...this.props} reduxStore={this.reduxStore} />
    }
  }
}
