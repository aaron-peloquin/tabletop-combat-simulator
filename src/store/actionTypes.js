/**
 * This is an object of Redux Actions. It mostly exists to
 *  ensure we do not create duplicate actions, but it can
 *  also be used as a list of our existing actions
*/
const actions = {
  CopyCreature: 'CopyCreature',
  DeleteAllCreatures: 'DeleteAllCreatures',
  DeleteCreature: 'DeleteCreature',
  SaveCreature: 'SaveCreature',
  SetEditCreature: 'SetEditCreature',
  ToggleSidebar: 'ToggleSidebar',
  RunSimulation: 'RunSimulation',
  UpdateEditCreature: 'UpdateEditCreature',
}

export default actions
