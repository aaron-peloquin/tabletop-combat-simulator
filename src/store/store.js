import {createStore, applyMiddleware, combineReducers} from "redux"
import {composeWithDevTools} from "redux-devtools-extension"
import thunkMiddleware from "redux-thunk"

/** Reducers */
import combat from "./reducer/combat"
import creatures from "./reducer/creatures"
import sideBar from "./reducer/sideBar"
import editing from "./reducer/editing"

/** Reducers */
const reducers = combineReducers({
  combat,
  creatures,
  editing,
  sideBar,
})

/**
 * Initializes the store
 * @param {obj} initialState the state our app starts with
 * @return {obj} the redux store
 */
export function initializeStore(initialState = initialState) {
  return createStore(reducers, initialState, composeWithDevTools(applyMiddleware(thunkMiddleware)))
}
