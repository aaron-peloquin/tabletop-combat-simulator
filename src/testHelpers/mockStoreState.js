export default {
  "creatures": [
    {
      "name": "Goplin",
      "cr": "1/4",
      "hash": "hivn3p0k",
      "description": "A stout monster",
      "defaultInitiative": 10,
      "armor": 11,
      "hp": 5,
      "hitDiceEquation": "1d20+1",
      "damageDiceEquation": "1d4"
    },
    {
      "name": "Elven Ranger",
      "cr": "1/2",
      "hash": "h3dxf5wl",
      "description": "A swift hunter",
      "defaultInitiative": 16,
      "armor": 13,
      "hp": 10,
      "hitDiceEquation": "1d20+8",
      "damageDiceEquation": "1d6+3"
    },
    {
      "name": "Knight",
      "cr": "1",
      "hash": "hee9p286",
      "description": "An honorable soldier",
      "defaultInitiative": 11,
      "armor": 18,
      "hp": 25,
      "hitDiceEquation": "1d20+4",
      "damageDiceEquation": "1d12+4"
    },
    {
      "name": "Lizardfolk",
      "cr": "3",
      "hash": "howx26rt",
      "description": "An ancient sneaky monster",
      "defaultInitiative": 14,
      "armor": 14,
      "hp": 15,
      "hitDiceEquation": "1d20+5",
      "damageDiceEquation": "1d10+2"
    }
  ],
  "sideBar": false,
}