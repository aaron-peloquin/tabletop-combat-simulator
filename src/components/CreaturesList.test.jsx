import {Provider} from "react-redux"
import {createShallow} from "@material-ui/core/test-utils"
import toJson from "enzyme-to-json"
import {initializeStore} from "../store/store"

import CreaturesList from "./CreaturesList"
import mockStoreState from "../testHelpers/mockStoreState"

describe("<CreaturesList />", ()=>{
  const renderShallowUntilComponent = createShallow({"untilSelector": "CreaturesList"})
  let store; let state; let Component
  beforeEach(async ()=>{
    store = initializeStore(mockStoreState)
    Component = renderShallowUntilComponent(<Provider store={store}><CreaturesList /></Provider>)
  })

  it("loads", () => {
    expect(typeof Component).toBe("object")
  })

  it("has a mui table", ()=>{
    expect(Component.find("WithStyles(Table)").length).toBe(1)
    expect(Component.find("WithStyles(TableHead)").length).toBe(1)
    expect(Component.find("WithStyles(TableBody)").length).toBe(1)
  })

  it("has a row for each creature", () => {
    state = store.getState()
    expect(state.creatures.length).toBe(Component.find("Connect(CreaturesTableRow)").length)
  })

  it("shows message when empty", () => {
    store = initializeStore({})
    Component = renderShallowUntilComponent(<Provider store={store}><CreaturesList /></Provider>)
    expect(Component.text()).toBe("No creatures exist yet")
  })

  it("Snapshots", () => {
    expect(toJson(Component)).toMatchSnapshot()
  })

  it("Snapshots without any creatures", () => {
    store = initializeStore({})
    Component = renderShallowUntilComponent(<Provider store={store}><CreaturesList /></Provider>)
    expect(toJson(Component)).toMatchSnapshot()
  })
})
