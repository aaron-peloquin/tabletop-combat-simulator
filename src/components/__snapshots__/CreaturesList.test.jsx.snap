// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CreaturesList /> Snapshots 1`] = `
<div>
  <p>
    Showing 
    4
     Creatures
  </p>
  <WithStyles(Table)>
    <WithStyles(TableHead)>
      <WithStyles(TableRow)>
        <WithStyles(TableCell)
          key="heading-"
        />
        <WithStyles(TableCell)
          key="heading-Name"
        >
          Name
        </WithStyles(TableCell)>
        <WithStyles(TableCell)
          key="heading-CR"
        >
          CR
        </WithStyles(TableCell)>
      </WithStyles(TableRow)>
    </WithStyles(TableHead)>
    <WithStyles(TableBody)>
      <Connect(CreaturesTableRow)
        creature={
          Object {
            "armor": 13,
            "damageDiceEquation": "1d6+3",
            "hash": "h3dxf5wl",
            "hitDiceEquation": "1d20+8",
            "hp": 10,
            "initiative": 16,
            "name": "Elven Ranger",
            "team": "b",
          }
        }
        key="creature-h3dxf5wl"
      />
      <Connect(CreaturesTableRow)
        creature={
          Object {
            "armor": 11,
            "cr": "1/4",
            "damageDiceEquation": "1d4",
            "hash": "hivn3p0k",
            "hitDiceEquation": "1d20+1",
            "hp": 5,
            "initiative": 10,
            "name": "Goplin",
            "team": "a",
          }
        }
        key="creature-hivn3p0k"
      />
      <Connect(CreaturesTableRow)
        creature={
          Object {
            "armor": 18,
            "damageDiceEquation": "1d12+4",
            "hash": "hee9p286",
            "hitDiceEquation": "1d20+4",
            "hp": 25,
            "initiative": 11,
            "name": "Knight",
            "team": "a",
          }
        }
        key="creature-hee9p286"
      />
      <Connect(CreaturesTableRow)
        creature={
          Object {
            "armor": 14,
            "damageDiceEquation": "1d10+2",
            "hash": "howx26rt",
            "hitDiceEquation": "1d20+5",
            "hp": 15,
            "initiative": 14,
            "name": "Lizardfolk",
            "team": "b",
          }
        }
        key="creature-howx26rt"
      />
    </WithStyles(TableBody)>
  </WithStyles(Table)>
</div>
`;

exports[`<CreaturesList /> Snapshots without any creatures 1`] = `
<p>
  No creatures exist yet
</p>
`;
